compress.go

```diff
@@ -74,14 +100,79 @@ func (c *Compressor) Next(prev, curr *lepton3.Frame) (uint8, []byte) {
 	return width, c.outBuf.Bytes()
 }
 
+// NewDecompressor creates a new Decompressor.
+func NewDecompressor() *Decompressor {
+	return &Decompressor{
+		rows:       lepton3.FrameRows,
+		cols:       lepton3.FrameCols,
+		pixelCount: lepton3.FrameRows * lepton3.FrameCols,
+		prevFrame:  new(lepton3.Frame),
+	}
+}
+
+// Decompressor is used to decompress successive CPTV frames. See the
+// Next() method.
+type Decompressor struct {
+	cols, rows, pixelCount int
+	prevFrame              *lepton3.Frame
+	deltas                 [lepton3.FrameRows][lepton3.FrameCols]int32
+}
+
+// ByteReaderReader combines io.Reader and io.ByteReader.
+type ByteReaderReader interface {
+	io.Reader
+	io.ByteReader
+}
+
+// Next reads of stream of bytes as a ByteReaderReader and
+// decompresses them using the bit width provided into the
+// lepton3.Frame provided.
+func (d *Decompressor) Next(bitWidth uint8, compressed ByteReaderReader, out *lepton3.Frame) error {
+	var v int32
+	err := binary.Read(compressed, binary.LittleEndian, &v)
+	if err != nil {
+		return err
+	}
+
+	unpacker := NewBitUnpacker(bitWidth, compressed)
+	d.deltas[0][0] = v
+	for i := 1; i < d.pixelCount; i++ {
+		y := i / lepton3.FrameCols
+		x := i % lepton3.FrameCols
+		// Deltas are "snaked" so work backwards through every second row.
+		if y%2 == 1 {
+			x = lepton3.FrameCols - x - 1
+		}
```

Comment:
I am guessing we would get a minor speed improvement if you move  `x := i % lepton3.FrameCols` into an `else` statement here so the modulus is not run twice in some loops.

Reply:
not really, because `x = lepton3.FrameCols - x - 1` relies on the the value of `x` calculated by `x := i % lepton3.FrameCols`.

I've just realised that the `if y%2 == 1` is silly though. That's an unnecessarily slow way of decided if a value is odd. `if y & 1 == 1` is faster.